java/source/infodynamics/measures/mixed/kraskov/MutualInfoCalculatorMultiVariateWithDiscreteKraskov.java

```diff
@@ -142,23 +248,71 @@ public void initialise(int dimensions, int base) {
 	 * 		default is {@link EuclideanUtils#NORM_MAX_NORM}.
 	 *  <li>{@link #PROP_NORMALISE} - whether to normalise the individual
 	 *      variables (true by default)</li>
+	 * 	<li>{@link #PROP_TIME_DIFF} - Time difference between source and
+   *     	destination (0 by default). Must be >= 0.</li>
+   *  <li>{@link #PROP_ADD_NOISE} -- a standard deviation for an amount of
+   *    random Gaussian noise to add to
+   *      each variable, to avoid having neighbourhoods with artificially
+   *      large counts. (We also accept "false" to indicate "0".)
+   *      The amount is added in after any normalisation,
+   *      so can be considered as a number of standard deviations of the data.
+   *      (Recommended by Kraskov. MILCA uses 1e-8; but adds in
+   *      a random amount of noise in [0,noiseLevel) ).
+   *      Default 1e-8 to match the noise order in MILCA toolkit.</li>
+   *  <li>{@link #PROP_DYN_CORR_EXCL_TIME} -- a dynamics exclusion time window,
+   *      also known as Theiler window (see Kantz and Schreiber);
+   *      default is 0 which means no dynamic exclusion window.</li>
 	 * </ul>
 	 * 
 	 * @param propertyName name of the property to set
 	 * @param propertyValue value to set on that property
 	 */
-	public void setProperty(String propertyName, String propertyValue) {
+	public void setProperty(String propertyName, String propertyValue)
+      throws Exception {
 		if (propertyName.equalsIgnoreCase(PROP_K)) {
 			k = Integer.parseInt(propertyValue);
 		} else if (propertyName.equalsIgnoreCase(PROP_NORM_TYPE)) {
-			normCalculator.setNormToUse(propertyValue);
+      normType = KdTree.validateNormType(propertyValue);
 		} else if (propertyName.equalsIgnoreCase(PROP_NORMALISE)) {
 			normalise = Boolean.parseBoolean(propertyValue);
+    } else if (propertyName.equalsIgnoreCase(PROP_DYN_CORR_EXCL_TIME)) {
+      dynCorrExclTime = Integer.parseInt(propertyValue);
+    } else if (propertyName.equalsIgnoreCase(PROP_ADD_NOISE)) {
+      if (propertyValue.equals("0") ||
+          propertyValue.equalsIgnoreCase("false")) {
+        addNoise = false;
+        noiseLevel = 0;
+      } else {
+        addNoise = true;
+        noiseLevel = Double.parseDouble(propertyValue);
+      }
+		} else if (propertyName.equalsIgnoreCase(PROP_TIME_DIFF)) {

```

Comment:
Actually we'll need to accept a time difference of below zero as well here. For continuous valued MI we could demand it be >= 0, because the user could still swap the variables in the method input to accommodate that. But for discrete to continuous they might genuinely want it the other way round because there's no way they can swap these variables in the method input. Can you allow negative time_diff property here as well please? This will also need to be supported in the implementation later.

Reply:
Good point. Just fixed this in commit https://github.com/pmediano/jidt/commit/d40024e1cd045844f14ca0ffa41ed386661c910d